File: frontend/src/types/uml.ts
export type Visibility = 'public' | 'private' | 'protected' | 'package';

export type RelationType = 'association' | 'inheritance' | 'aggregation' | 'composition';

export interface UMLAttribute {
  id: string;
  name: string;
  type: string;
  visibility: Visibility;
  multiplicity: string;
}

export interface UMLMethodParameter {
  id: string;
  name: string;
  type: string;
}

export interface UMLMethod {
  id: string;
  name: string;
  returnType: string;
  visibility: Visibility;
  parameters: UMLMethodParameter[];
}

export interface UMLClass {
  id: string;
  name: string;
  visibility: Visibility;
  isAbstract: boolean;
  attributes: UMLAttribute[];
  methods: UMLMethod[];
}

export interface UMLRelation {
  id: string;
  source: string;
  target: string;
  type: RelationType;
  cardinality: string;
}

export interface UMLModel {
  classes: UMLClass[];
  relations: UMLRelation[];
}

export interface BackendSavedModel {
  id: string;
  name: string;
  model: UMLModel;
  createdAt: string;
}

File: backend/app/models/uml.py
from typing import List, Literal
from pydantic import BaseModel, Field

Visibility = Literal['public', 'private', 'protected', 'package']
RelationType = Literal['association', 'inheritance', 'aggregation', 'composition']

class UMLAttribute(BaseModel):
    id: str
    name: str
    type: str
    visibility: Visibility = 'public'
    multiplicity: str = '1'

class UMLMethodParameter(BaseModel):
    id: str
    name: str
    type: str

class UMLMethod(BaseModel):
    id: str
    name: str
    returnType: str = 'void'
    visibility: Visibility = 'public'
    parameters: List[UMLMethodParameter] = Field(default_factory=list)

class UMLClass(BaseModel):
    id: str
    name: str
    visibility: Visibility = 'public'
    isAbstract: bool = False
    attributes: List[UMLAttribute] = Field(default_factory=list)
    methods: List[UMLMethod] = Field(default_factory=list)

class UMLRelation(BaseModel):
    id: str
    source: str
    target: str
    type: RelationType
    cardinality: str = '1'

class UMLModel(BaseModel):
    classes: List[UMLClass] = Field(default_factory=list)
    relations: List[UMLRelation] = Field(default_factory=list)
